File: sample-apps/python-auto-instrumentation-sample-app/random_metrics.py
import random
import threading
import time
import os

from config import *
from opentelemetry.metrics import CallbackOptions, Observation
from opentelemetry import metrics

cfg = create_config('config.yaml')

common_attributes = { 'signal': 'metric', 'language': 'python-auto-instrumentation', 'metricType': 'random' }

meter = metrics.get_meter(__name__)

# Starts the callback for cpu usage
def cpu_usage_callback(options: CallbackOptions):
    min = 0
    max = cfg['RandomCpuUsageUpperBound']
    cpu_usage = Observation(value=random.randint(min, max), attributes=common_attributes)
    print('CPU Usage asked by SDK')
    yield cpu_usage

# Starts the callback for heap size
def heap_size_callback(options: CallbackOptions):
    min = 0
    max = cfg['RandomTotalHeapSizeUpperBound']
    total_heap_size = Observation(value=random.randint(min, max), attributes=common_attributes)
    print("Heapsize asked by SDK")
    yield total_heap_size

# This is the random metric collector class
class RandomMetricCollector():
    
    # Init registers 4 different metrics
    def __init__(self):
        testingId = ""
        if (os.environ.get("INSTANCE_ID")):
            testingId = "_" + os.environ["INSTANCE_ID"]

        self.time_alive=meter.create_counter(
            name="time_alive" + testingId,
            description="Total amount of time that the application has been alive",
            unit='ms'
        )
        self.cpu_usage=meter.create_observable_gauge(
            name="cpu_usage" + testingId,
            callbacks=[cpu_usage_callback],
            description="Cpu usage percent",
            unit='1'
        )
        self.threads_active=meter.create_up_down_counter(
            name="threads_active" + testingId,
            description="The total number of threads active",
            unit='1'
        )
        self.total_heap_size=meter.create_observable_gauge(
            name="total_heap_size" + testingId,
            callbacks=[heap_size_callback],
            description="The current total heap size",
            unit='By'
        )
        self.thread_bool = True
        self.thread_count = 0

    # Adds one to the time alive counter
    def update_time_alive(self, cfg=None):
        self.time_alive.add(cfg['RandomTimeAliveIncrementer'], attributes=common_attributes)
    
    # Updates the currently active threads based on its current bounds.
    def update_threads_active(self, cfg=None):
        if self.thread_bool:
            if self.thread_count < cfg['RandomThreadsActiveUpperBound']:
                self.threads_active.add(1, attributes=common_attributes)
                self.thread_count += 1
            else:
                self.thread_bool = False
                self.thread_count -= 1
        
        else:
            if self.thread_count > 0 :
                self.threads_active.add(-1, attributes=common_attributes)
                self.thread_count -= 1
            else:
                self.thread_bool = True
                self.thread_count += 1

    # This function registers the metrics client 
    def register_metrics_client(self,cfg=None):
        
        def update(self, cfg):
            while True:
                self.update_time_alive(cfg)
                self.update_threads_active(cfg)
                print("updating time alive & active threads...")
                time.sleep(cfg['TimeInterval'])

        update_thread = threading.Thread(target=update, args=(self, cfg,), daemon=True)
        update_thread.start()

File: sample-apps/javascript-sample-app/random-metrics.js
/*
 * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License").
 * You may not use this file except in compliance with the License.
 * A copy of the License is located at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * or in the "license" file accompanying this file. This file is distributed
 * on an "AS IS'" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 * express or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 *
 */

'use strict'

const sdk = require("./nodeSDK");
const api = require('@opentelemetry/api');

// config
const create_cfg = require('./config');
const cfg = create_cfg.create_config('./config.yaml');

const TIME_ALIVE_METRIC = 'time_alive';
const CPU_USAGE_METRIC = 'cpu_usage';
const THREADS_ACTIVE_METRIC = 'threads_active';
const HEAP_SIZE_METRIC = 'total_heap_size';

const common_attributes = { signal: 'metric',  language: 'javascript', metricType: 'random' };

let threadCount = 0;
let cpuUsage = 0;
let totalHeapSize = 0;

// acquire meter 
const meter = api.metrics.getMeter('js-sample-app-meter');
var testingId = "";
if (process.env.INSTANCE_ID) {
    testingId = "_" + process.env.INSTANCE_ID
}

// counter metric
const timeAliveMetric = meter.createCounter(TIME_ALIVE_METRIC + testingId, {
    description: 'Total amount of time that the application has been alive',
    unit: 's'
});

// updown counter metric
const threadsActiveMetric = meter.createUpDownCounter(THREADS_ACTIVE_METRIC + testingId, {
    description: 'The total number of threads active',
    unit:'1'
});

// observable gauge metric
const cpuUsageMetric = meter.createObservableGauge(CPU_USAGE_METRIC + testingId, {
    description: 'Cpu usage percent',
    unit: '1'
});
cpuUsageMetric.addCallback((measurement) => {measurement.observe(cpuUsage, common_attributes)});

// observable updown counter metric
const totalHeapSizeMetric = meter.createObservableUpDownCounter(HEAP_SIZE_METRIC + testingId, {
    description: 'The current total heap size',
    unit:'1'
});
totalHeapSizeMetric.addCallback((measurement) => {measurement.observe(totalHeapSize, common_attributes)});

// updates observable gauge
function updateCpuUsageMetric() {
    cpuUsage = Math.random() * cfg.RandomCpuUsageUpperBound;
}

// updates updown counter
function updateSizeMetric() {
    totalHeapSize += Math.random() * cfg.RandomTotalHeapSizeUpperBound;
}

// updates counter
function updateTimeAlive() {
    timeAliveMetric.add(cfg.RandomTimeAliveIncrementer, common_attributes);
}

// updates updown counter metric
function updateThreadsActive() {
    let min = -1;
    let activeThreads = Math.floor(Math.random() * (cfg.RandomThreadsActiveUpperBound - min + 1) + min);

    if (threadCount < activeThreads) {
        threadsActiveMetric.add(1, common_attributes);
        threadCount++;
    }
    else {
        threadsActiveMetric.add(-1, common_attributes);
        threadCount--;
    }
}

module.exports = { updateTimeAlive, updateThreadsActive, updateCpuUsageMetric, updateSizeMetric };
